core/src/test/java/io/grpc/DeadlineTest.java

```diff
@@ -45,62 +43,112 @@
 
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.junit.runners.JUnit4;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
 
+import java.util.Arrays;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 /**
  * Tests for {@link Context}.
  */
-@RunWith(JUnit4.class)
+@RunWith(Parameterized.class)
 public class DeadlineTest {
+  /** Ticker epochs to vary testing. */
+  @Parameters
+  public static Iterable<Object[]> data() {
+    return Arrays.asList(new Object[][] {
+      // MAX_VALUE / 2 is important because the signs are generally the same for past and future
+      // deadlines.
+      {Long.MAX_VALUE / 2}, {0}, {Long.MAX_VALUE}, {Long.MIN_VALUE}
+    });
+  }
+
+  private FakeTicker ticker = new FakeTicker();
 
-  // Allowed inaccuracy when comparing the remaining time of a deadline.
-  private final long maxDelta = TimeUnit.MILLISECONDS.toNanos(20);
+  public DeadlineTest(long epoch) {
+    ticker.reset(epoch);
+  }
+
+  @Test
+  public void defaultTickerIsSystemTicker() {
+    Deadline d = Deadline.after(0, TimeUnit.SECONDS);
+    ticker.reset(System.nanoTime());
+    Deadline reference = Deadline.after(0, TimeUnit.SECONDS, ticker);
+    // Allow inaccuracy to account for system time advancing during test.
+    assertAbout(deadline()).that(d).isWithin(20, TimeUnit.MILLISECONDS).of(reference);
+  }
+
+  @Test
+  public void timeCanOverflow() {
+    ticker.reset(Long.MAX_VALUE);
+    Deadline d = Deadline.after(10, TimeUnit.DAYS, ticker);
+    assertEquals(10, d.timeRemaining(TimeUnit.DAYS));
+    assertTrue(Deadline.after(0, TimeUnit.DAYS, ticker).isBefore(d));
+    assertFalse(d.isExpired());
+
+    ticker.increment(10, TimeUnit.DAYS);
+    assertTrue(d.isExpired());
+  }
+
+  @Test
+  public void timeCanUnderflow() {
+    ticker.reset(Long.MIN_VALUE);
+    Deadline d = Deadline.after(-10, TimeUnit.DAYS, ticker);
+    assertEquals(-10, d.timeRemaining(TimeUnit.DAYS));
+    assertTrue(d.isBefore(Deadline.after(0, TimeUnit.DAYS, ticker)));
+    assertTrue(d.isExpired());
+  }
+
+  @Test
+  public void deadlineClamps() {
+    Deadline d = Deadline.after(-300 * 365, TimeUnit.DAYS, ticker);
+    Deadline d2 = Deadline.after(300 * 365, TimeUnit.DAYS, ticker);
+    assertTrue(d.isBefore(d2));
+
+    Deadline d3 = Deadline.after(-200 * 365, TimeUnit.DAYS, ticker);
+    // d and d3 are equal
+    assertFalse(d.isBefore(d3));
```

Comment:
I didn't because I don't quite know what's going to be happening with compareTo vs isBefore and the DeadlineSubject. Basically the gains by your previous commit are all but gone (only `defaultTickerIsSystemTicker()` is using it). I think there's plenty of reasons not to remove your additions, but I also didn't want a false sense of necessity to cloud the discussion.